crates/red_knot_python_semantic/src/types.rs

```diff
@@ -782,18 +817,40 @@ impl<'db> Type<'db> {
         // TODO: The following is a workaround that is required to unify the two different versions
         // of `NoneType` and `NoDefaultType` in typeshed. This should not be required anymore once
         // we understand `sys.version_info` branches.
-        self == other
-            || matches!((self, other),
-                (
-                    Type::Instance(InstanceType { class: self_class }),
-                    Type::Instance(InstanceType { class: target_class })
-                )
-                if {
-                    let self_known = self_class.known(db);
-                    matches!(self_known, Some(KnownClass::NoneType | KnownClass::NoDefaultType))
-                        && self_known == target_class.known(db)
-                }
+        match (self, other) {
+            (
+                Type::Instance(InstanceType { class: self_class }),
+                Type::Instance(InstanceType {
+                    class: target_class,
+                }),
+            ) if {
+                let self_known = self_class.known(db);
+                matches!(
+                    self_known,
+                    Some(KnownClass::NoneType | KnownClass::NoDefaultType)
+                ) && self_known == target_class.known(db)
+            } =>
+            {
+                true
+            }
```

Comment:
this gets a little hard to read 😄 how about writing it like this?

```suggestion
            (Type::Instance(self_instance), Type::Instance(target_instance)) => {
                let Some(self_known_class) = self_instance.class.known(db) else {
                    return false;
                };
                if !matches!(
                    self_known_class,
                    KnownClass::NoneType | KnownClass::NoDefaultType
                ) {
                    return false;
                }
                Some(self_known_class) == target_instance.class.known(db)
            }
```

Reply:
We have to be careful here; these `return false` in the suggested change are not correct. They would have to be `return self == other`, and I'm not sure we want to duplicate that condition in various places.

Reply:
I had also considered handling each of the special cases with a separate `if` statement + early return.  Then it would possibly be clearer that `self == other` is the fall-back / general case.  But we had other large `match` blocks elsewhere that made me think that might be preferred house style.  I will take a stab at the early return pattern and see if it's more readable.

Reply:
> I will take a stab at the early return pattern and see if it's more readable.

Now using early return.  lmkwyt